src/bits32/paging.rs

```diff
@@ -0,0 +1,1431 @@
+pub struct Page([u8; BASE_PAGE_SIZE]);
+
+/// A type wrapping a large page with a 4 MiB buffer.
+pub struct LargePage([u8; LARGE_PAGE_SIZE]);
+
+/// Mask to find the physical address of an entry in a page-table.
+const ADDRESS_MASK: u32 = !0xfff;
+
+/// Page tables have 512 = 4096 / 32 entries.
+pub const PAGE_SIZE_ENTRIES: usize = 1024;
+
+/// A page directory.
+pub type PD = [PDEntry; PAGE_SIZE_ENTRIES];
+
+/// A page table.
+pub type PT = [PTEntry; PAGE_SIZE_ENTRIES];
+
+/// Given virtual address calculate corresponding entry in PD.
+#[inline]
+pub fn pd_index(addr: VAddr) -> usize {
+    ((addr >> 22usize) & 0b1111111111) as usize
+}
+
+/// Given virtual address calculate corresponding entry in PT.
+#[inline]
+pub fn pt_index(addr: VAddr) -> usize {
+    ((addr >> 12usize) & 0b1111111111) as usize
+}
+
+bitflags! {
+    /// PD configuration bits description.
+    #[repr(transparent)]
+    pub struct PDFlags: u32 {
+        /// Present; must be 1 to map a 4-MByte page.
+        const P       = bit!(0);
+        /// Read/write; if 0, writes may not be allowed to the 4-MByte page referenced by this entry.
+        const RW      = bit!(1);
+        /// User/supervisor; if 0, user-mode accesses are not allowed to the 4-MByte page referenced by this entry.
+        const US      = bit!(2);
+        /// Page-level write-through.
+        const PWT     = bit!(3);
+        /// Page-level cache disable.
+        const PCD     = bit!(4);
+        /// Accessed; indicates whether software has accessed the 4-MByte page referenced by this entry.
+        const A       = bit!(5);
+        /// Dirty; indicates whether software has written to the 4-MByte page referenced by this entry.
+        const D       = bit!(6);
+        /// Page size; if set this entry maps a 4-MByte page; otherwise, this entry references a page directory.
+        const PS      = bit!(7);
+        /// Global; if CR4.PGE = 1, determines whether the translation is global; ignored otherwise.
+        const G       = bit!(8);
+        /// If the PAT is supported, indirectly determines the memory type used to access the 4-MByte page referenced by this entry;
+        /// otherwise, reserved (must be 0)
+        const PAT     = bit!(12);
+    }
+}
+
+/// A PD Entry consists of an address and a bunch of flags.
+#[repr(transparent)]
+#[derive(Clone, Copy)]
+pub struct PDEntry(pub u32);
+
+impl fmt::Debug for PDEntry {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "PDEntry {{ {:#x}, {:?} }}", self.address(), self.flags())
+    }
+}
+
+impl PDEntry {
+    /// Creates a new PDEntry.
+    ///
+    /// # Arguments
+    ///
+    ///  * `pt` - The physical address of the page table.
+    ///  * `flags`- Additional flags for the entry.
+    pub fn new(pt: PAddr, flags: PDFlags) -> PDEntry {
+        let pt_val = pt & ADDRESS_MASK;
+        assert!(pt_val == pt.into());
+        assert!(pt % BASE_PAGE_SIZE == 0);
+        PDEntry(pt_val | flags.bits)
```

Comment:
![image](https://user-images.githubusercontent.com/77098789/121347397-1b7bac80-c962-11eb-988e-47b65f80fb19.png)
I noticed that using PDEntry for 4MiB page has a problem.
If PS = 1, bit manipulation is required to follow to this specification, but PAddr can't represent 39:32 bits.

Reply:
Agreed, I think that on 32bit systems the PAddr should probably be 64bits. e.g. ACPI might give you physical address ranges above 2^32, similarly PCI devices can be above the 4G limit. I haven't programmed for rust 32bit enough to know if you can just change PAddr to PAddr(u64) or if you need to split it into two low/high  variables, both 32bit in size?

Reply:
It seems to be a feature to support PSE-36 and PSE-40.
I will drop it for now, but I might add support for PSE-* or PAE in the future if needed. Also PAddr is a transparent struct, so I think it's better to use low/high variables rather than to resize. (Or add ExtendedPAddr...)